test-junit/jvm/src/main/scala/zio/test/junit/ZTestJUnitRunner.scala

```diff
@@ -0,0 +1,143 @@
+package zio.test.junit
+
+import org.junit.runner.manipulation.{ Filter, Filterable }
+import org.junit.runner.notification.{ Failure, RunNotifier }
+import org.junit.runner.{ Description, RunWith, Runner }
+import zio.ZIO.effectTotal
+import zio.clock.Clock
+import zio.test.FailureRenderer.FailureMessage.Message
+import zio.test.Spec.{ SuiteCase, TestCase }
+import zio.test.TestFailure.{ Assertion, Runtime }
+import zio.test.TestSuccess.{ Ignored, Succeeded }
+import zio.test.environment.TestEnvironment
+import zio.test.{ ExecutedSpec, _ }
+import zio.{ DefaultRuntime, URIO, ZIO }
+
+import scala.util.Try
+
+class ZTestJUnitRunner(klass: Class[_]) extends Runner with Filterable with DefaultRuntime {
+
+  private val className = klass.getName.stripSuffix("$")
+
+  private lazy val spec: AbstractRunnableSpec = {
+    Try(loadObject())
+      .getOrElse(
+        klass
+          .newInstance()
+          .asInstanceOf[AbstractRunnableSpec]
+      )
+  }
+
+  private def loadObject(): AbstractRunnableSpec = {
+    import org.portablescala.reflect._
+    val fqn = className + "$"
+    Reflect
+      .lookupLoadableModuleClass(fqn, klass.getClassLoader)
+      .getOrElse(throw new ClassNotFoundException("failed to load object: " + fqn))
+      .loadModule()
+      .asInstanceOf[AbstractRunnableSpec]
+  }
+
+  private var filter = Filter.ALL
+
+  lazy val getDescription: Description = {
+    val description = Description.createSuiteDescription(className)
+    def traverse[R, E, L, S](
+      spec: ZSpec[R, E, L, S],
+      description: Description,
+      path: Vector[String] = Vector.empty
+    ): URIO[R, Unit] =
+      spec.caseValue match {
+        case SuiteCase(label, specs, _) =>
+          val suiteDesc = Description.createSuiteDescription(label.toString, path.mkString(":"))
+          effectTotal(description.addChild(suiteDesc)) *>
+            specs
+              .flatMap(ZIO.foreach(_)(traverse(_, suiteDesc, path :+ label.toString)))
+              .catchAll(_ => ZIO.unit)
+              .unit
+        case TestCase(label, _) =>
+          effectTotal(description.addChild(testDescription(label.toString, path)))
+      }
+    unsafeRun(
+      traverse(filteredSpec, description)
+      // we expect to be able to extract structure without the environment
+        .provide(null.asInstanceOf[spec.Environment])
```

Comment:
@adamgfraser When traversing the structure of the spec, I couldn't find a way to get the environment from the spec's runner. Just providing `null` for now, but this is ugly and may fail if people do rely on the environment for structure... Any ideas?

Reply:
Ok, I found a way to get the environment from `TestExecutor`, but I had promote it to trait and expose the managed environment.